### DISCUSSION TITLE
The state of PSPGL

### DISCUSSION THREAD

**Discussion by tpimh** [pspdev/pspdev#72] [General]
We currently have two pspGL repos: [pspdev/pspgl](https://github.com/pspdev/pspgl) (active) and [pspdev/pspgl-1](https://github.com/pspdev/pspgl-1) (unmaintained, but as long commit history). The code in them is mostly identical, but has some conflicts in history. For example, here two different people are doing the same thing some years apart from each other:

1. @jsgf fixes hardcoded error value on Jul 26, 2006: https://github.com/pspdev/pspgl-1/commit/3229820e3f20e9c6340c6ffbd79597f84c2583c5
2. @libcg does the same on Jan 31, 2012: https://github.com/pspdev/pspgl/commit/8bab4c6c5fe31f44e8a47d5d45bdbf774bed83c1

Here is my attempt to merge the two without losing anything: [tpimh/pspgl](https://github.com/tpimh/pspgl) (rewriting history a tiny little bit, but I believe it's for the best).

Here is the project history:
1. The project was originally started by Holger Waechtler in 2005. It began with a total of ~6.5k lines in 100 files. Absolute legend.
2. There was no version control until someone aliased as `stefan` put the code to (presumably) svn.
3. `tyranid`, `shine`, `mrbrown` and `jim` also made a couple commits additionally to 132 commits by the original author.
4.  Jeremy Fitzhardinge @jsgf joined in the late 2005, and did a lot of work in 2005-2006. He was initially using Mercurial for version control and maintaining an svn mirror, and later converted it to git in 2009.
5. First fork was made by @ooPo in 2011. The code was copied from svn repo (with some changes and additions), so the history was not preserved.
6. The main pspdev repo is based on that fork, more contributions were added to it in 2012 (@libcg), 2020 (@iyenal, @dbeef, @sharkwouter, @carstene1ns) and 2022 (@rofl0r).
7. There is [another fork](https://gitlab.com/HaydenKow/pspgl/) by @mrneo240 which is based on Jeremy's git conversion. It had activity in 2019-2021.

*Sorry if I forgot someone, feel free to correct me.*

Also Jeremy wrote some nice documentation, which which was preserved and can be viewed in the Wayback Machine. I made a manual conversion to markdown: [tpimh/pspgl-docs](https://github.com/tpimh/pspgl-docs). It should most likely be published on [pspdev.github.io](https://pspdev.github.io/) for easier access.

What shall we do? I suggest overwriting the contents of [pspdev/pspgl](https://github.com/pspdev/pspgl) with my restoration, then deleting [pspdev/pspgl-1](https://github.com/pspdev/pspgl-1) as it's no longer useful, only confusing.

Additional question: is it PSPGL, pspgl or pspGL? Nobody knows...

**Comment by fjtrujy** [comment]
This "way" to do things has been happening in "pspdev" in several repos.
Instead of forking the original repo, they were creating a copy of the repo, happen something similar with `newlib` and `gcc`.

My recommendation to follow is something else:
- Remove the current `pspdev/pspgl`
- Then rename the `pspdev/pspgl-1` to `pspdev/pspgl`
- Finally to override master content with the content your [tpimh/pspgl](https://github.com/tpimh/pspgl)
- As a bonus configure properly in the new `pspdev/pspgl` the documentation you mentioned.

I think I have grants to do all these things, if not I will ask @ooPo 

Cheeers

**Reply by tpimh** [reply]
What is the benefit of removing `pspdev/pspgl` (24⭐) and keeping `pspdev/pspgl-1` (1⭐)? Is it because it's a fork?

**Reply by tpimh** [reply]
If it's this, then it's possible to [attach a fork](https://support.github.com/request/fork) to the correct repo.

**Reply by fjtrujy** [reply]
Yes, keep the “real history”, which is the fork (if the fork is the real source of the code).
I don’t care that much about stars if we are speaking 24 vs 1, if all the pspdev put starts to the new one quickly we will have even more…

**Reply by fjtrujy** [reply]
> If it's this, then it's possible to [attach a fork](https://support.github.com/request/fork) to the correct repo.

I wasn’t aware that this was possible to do, super!
So then let’s attach the fork, override the history with your changes, and remove gl1 repo.

**Reply by tpimh** [reply]
In my opinion, the correct order should be:
1. Override the history
2. Attach the fork
3. Remove old repo

Right now there are no common commits in `pspdev/pspgl` and `jsgf/pspgl`.

**Reply by fjtrujy** [reply]
Nice!
I'm totally fine with this solution

**Reply by rofl0r** [reply]
before you start to overwrite history, which will break all existing clones, could you first provide a diff with the changes between the repos ? i personally don't think the value of the real history is so important it justifies modifying a repo that's in use by potentially hundreds of people.
if both repos share a common ancestor it could be made into a branch "real-history" for people interested into that sort of thing.

**Reply by tpimh** [reply]
This is a really good point. I guess just adding a new branch is better than completely overwriting the history.

**Reply by fjtrujy** [reply]
I think that keeping past mistakes is not a good idea.

So if you want to keep the previous `sha`s to avoid affecting "hundreds of users" (I bet there is not more than 10 xD, and most probably they are part of the `pspdev` community), you can rename the current branch to "legacy" and then rewrite the history of the current master/main.
I think that having maser/main branch as the "default" one, with the desired state is the best to do.

**Reply by rofl0r** [reply]
fwiw, i did a similar thing with my gnuboy repo - the original svn history is lost for eternity, but the original author provided me with a set of release tarballs and snapshots he had stored on his disk, so i first created "history" branch, then made a second branch based on that that fixed all build errors for the respective history with modern compilers, so the history could be used for bisecting. https://github.com/rofl0r/gnuboy/branches . i didn't modify master so people having a clone would not run into errors on pull.

**Reply by tpimh** [reply]
In this case, let's do this:

1. Create a new branch in `pspdev/pspgl` (for example `dev`) and keep `master` as it is
2. Overwrite the history of just this branch
3. Make it the [default branch](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/changing-the-default-branch)
4. Request to [attach a fork](https://support.github.com/request/fork)
5. Delete `pspdev/pspgl-1`

This way we will restore the correct commit history, keep the stars and won't break anything for anyone.

**Comment by mrneo240** [comment]
Every Readme says pspGL as name

**Reply by tpimh** [reply]
Docs: PSPGL (35 times)
Readme: pspGL (1 time)

**Comment by jsgf** [comment]
Yeesh 
![image](https://github.com/pspdev/pspdev/assets/147966/d80b21ec-b2ff-4ee1-8085-e9cf678bfff8)

I'm happy with whatever makes this most useful. I'm glad those docs are preserved and still helpful to people.

**Comment by fjtrujy** [comment]
To attach a fork is not something direct, you need to request it via email.
I honestly don't think it is a major issue to "break" those ones that have the repo previously downloaded.
I don't expect more than a few of US with the repo downloaded...

**Reply by rofl0r** [reply]
tbh i don't really care how it is done as long as the new master has identical content to the old one. 
so optimally before it is swapped out we have both available for a few days so they can be compared.
since the repo is slim it can be recloned quickly. when you otoh rebase a huge repo like linux kernel, people will be quite pissed if they have to redownload gigabytes of stuff.

**Reply by fjtrujy** [reply]
> tbh i don't really care how it is done as long as the new master has identical content to the old one.

I actually don't really care about those users, I do care about myself maintaining the repo.
Basically, those users will try do:
- `git fetch` will work
- `git pull` will fail and it will tell you that your branch has been diverged.
- `git reset --hard origin/master` will solve the issue.

We are going to have forever a useless branch called "master", just for trying to avoid potential clients that have the `pspgl` downloaded. but let's see honestly, how many users do you expect to have that repo cloned? And how many of them aren't part of `pspdev`?

I still think it is a mistake

**Reply by rofl0r** [reply]
it seems you misunderstood what i meant with content. i mean that foo.c will be identical in current repo and rewritten retro. not talking about commits.
i also said that pspgl is such a tiny repo that it doesnt matter when those few users have to clone it again. please read my post again with that in mind. it basically agrees with you entirely.

**Reply by sharkwouter** [reply]
Ah, you don't want the file structure to change because of the transition to the newer code. That makes sense. I would also vote in favor or rewriting the git history to be able to merge these efforts.

**Reply by rofl0r** [reply]
right. i think before making the new repo with the rewritten history the official pspgl repo a recursive diff should be made so we dont actually have a code regression due to a mistaken omission of commits or anything else. apart from that i basically only wanted to express that  i'm fine with however this ends up being implemented and don't want to stand in the way of the most pragmatic solution.

**Reply by tpimh** [reply]
The only reason to overwrite the contents of the old repo instead of creating a new fork is to keep the stars and maybe some other metadata.

**Reply by sharkwouter** [reply]
I think it's nice to have pspdev/pspgl. Easy to find.
